### PR TITLE
Single Logout Service support

### PR DESCRIPTION
**Pull request by peppelinux**
SLO support [HTTP-REDIRECT and HTTP-POST]

In addition:
- urls.py refactored, less code same features
- some code and function decoupled in views.py (more would come in the future)
- idp example setup with SLO endpoint
- gettext lazy localizazion in view (internazionalization support)

### CODE REVIEW

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py]: This function isn't called anywhere else. I think it might be better to put its code directly into saml_session_request.

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:62]: I'm not 100% sure that all implementations send the SigAlg and Signature with the initial request. Also, since the original verification implementation has been replaced by verify_request_signature, thos code is probably redundant and cna be deleted

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py]: Do we want to give this a more descriptive name? It isn't obvious what it does (moving POST/GET into session) until later inspection. Also, maybe when calling it we could say request = saml_session_request(request)? I know that functionally it makes no difference, but could help readability

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:178]: If it doesn't have a processor, it should not have gotten to this stage. I think this check is unnecessary

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:179]: As a side note, this is where I'd like to put in the redirect for the user agreement screen. It wouldn't interrupt the logical code flow, wouldn't have too many points of failure, and would act simply and logically.

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:187]: I don't think it's necessary to have the template as a variable. As per the way Django works, it's easy to override templates, so I dont see how this would be used

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:226]: Correct me if I'm wrong here. When we call this method, we're not doing anything with what it returns. As such, any error handling that it returns will be discarded, so it will fail to catch errors. If we want to move verifying signature, setting SP, etc, into methods, we need to take another approach to error checking. Perhaps we could wrap all these method calls in an exception handler, and have them raise exceptions instead of directly using the handle error view?

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:232]: See my above comment. Also, this should be renamed to check_access()

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:111]: For this and the next 2 methods, the error handling won't work because we're not doing anything with the return value. I'd say we should raise the exceptions in the methods, and then wrap all the calls to the methods in an error handler that would then return the handle_error view

File: djangosaml2idp/views.py
Comment (askvortsov1) [inline djangosaml2idp/views.py:131]: Why are you declaring the variable, and not just checking if not req_info.signature_check(req_info.xmlstr)?

File: djangosaml2idp/views.py
Comment (askvortsov1) [reply djangosaml2idp/views.py:187]: I would return this part to the way it was before. Templates can be overriden without changing the template location (https://docs.djangoproject.com/en/2.1/howto/overriding-templates/)

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
I've left comments on individual files. It would probably also be a good idea to add documentation about this.

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py]: I agree, saml_session_request is the place to get all the things there. Done

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:62]: I will remove once all the tests were made, assigned to me, I agree.

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py]: Yes, modify it how you prefer. This morning I thought the same thing!

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:178]: Legacy code, I did not tested the multi_factor_auth yet. Feel free to adjust as you think

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:179]: Ok

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:187]: I know but I do not love constants in the code. No problem for me if you want revert this change, feel free. I'll remember this decision.

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:226]: You're right, this method was created just as an delegation interface. Otherwise today I've tested pySAML2 internals about sign verification and I hae good news. Bytheway I'd like to hold another validation in our code, to avoids problem for future pySAML2 bugs or implementation errors

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:232]: Ok

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:111]: Ok, If I remeber well these could be legacy code that I adapted with internazionalization, nothing else. I agree

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py:131]: legacy code, verified_ok comes from it. I Agree, it's strange but I thought the same thing :)

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py]: Done here
https://github.com/peppelinux/djangosaml2idp/commit/1975bbdbdae7f6acbc61217f16e86864ab590632#diff-d320bc58b091d46b1ff068902e6a074dR40

File: djangosaml2idp/views.py
Comment (peppelinux) [reply djangosaml2idp/views.py]: Excuse me, I didn't want to delegate this to you, I can do it soon as possibile I mean, letting you to feel free to do changements!
This is valid also for next comments!!!

File: djangosaml2idp/views.py
Comment (askvortsov1) [reply djangosaml2idp/views.py:187]: As a compromise, we could assign the template to a variable on the previous line and take it out of the params. That way it wouldn't be a loose constant, but it also wouldn't be in the params for no reason.

File: djangosaml2idp/views.py
Comment (askvortsov1) [reply djangosaml2idp/views.py]: It's no problem, I'll have made the adjustments shortly

File: djangosaml2idp/views.py
Comment (askvortsov1) [reply djangosaml2idp/views.py:178]: Fixed

File: djangosaml2idp/views.py
Comment (askvortsov1) [reply djangosaml2idp/views.py:62]: Fixed

File: djangosaml2idp/views.py
Comment (askvortsov1) [reply djangosaml2idp/views.py:131]: Fixed

File: djangosaml2idp/views.py
Comment (askvortsov1) [reply djangosaml2idp/views.py:187]: Fixed

**Comment by peppelinux** [comment]
I previously merged It in to master,  now we Will get some conflicts when you'll merge  SLO. I'll can do separate branch for metadata download, tell me how to manage these desynced branches for a cost-less merge. Will you patch slo code style by yourself or there would be some cherry-picking? It's time to decide a milestone :)

**Comment by askvortsov1** [comment]
Hi, sorry about the confusion here. I'm going through right now and putting in all fixes as per previous discussion. I'll have that done within about 20 minutes at most, then merge SLO into master (I didnt want to merge directly into master because some things would cause critical errors if merged into master without immediately fixing. We can resume as normal after that.

**Comment by askvortsov1** [comment]
Ok, I've finished that up. In regards to deciding a milestone, do you mean setting out a formal organization of features we want to add, as well as how we want to add them? Sorry, I'm a bit new to open source so I'm not yet familiar with all conventions.

**Comment by peppelinux** [comment]
The same as actual roadmap, Will decide features opening issues and which features Will have the next release. We'll a release candidate (rc) without unit test, then a release production ready with the unit test. So, we could decide to do the next rc with agreement screen and metadata validator (as issues describes), then we'll put unit testa and code coverage to produce a real production ready release.

An important thing Is that you refactored the code, so each features would be developed in a separate branch and should be developed without touching so much the core code.

**Comment by askvortsov1** [comment]
We don't have access to manage the main parent branch though, or to release anything. Should we set the roadmap up on my branch then? Or wait for the primary maintainer to be more communicative?

On that note, we need to consider what to do if @mhindery isn't able to respond to our modifications. As seen in https://github.com/OTA-Insight/djangosaml2idp/issues/28, there are bugs in the old version that have been resolved with mine and @peppelinux's modifications. We've also added a heap of new features, and are relatively actively working on adding even more. How are these situations generally resolved?

As to developing major features in their own branches, I completely agree. Is the workflow I followed in https://github.com/askvortsov1/djangosaml2idp/pull/4 close to what you're thinking of?

**Comment by peppelinux** [comment]
If @mhindery is not interested to this work we should create an organization with a developer team, at the moment we can wait for @mhindery interactions, if he will introduce us in the dev team, and permits to manage repository or if he prefers to carry the development with relaxed times. The most important thing is to decide which features should have every single release, to be accurate in our goals. I pushed to your repo because you're very active and this is a very good thing for project's healt!

We should decide with @mhindery the features and the release candidate name, creating a document (it could be an Issue istance as well), this way we should then close then all the current related issues. For the first release we should activate travis and code coverage to give to the community production-ready assurances.

Now I'm on your latest commits.

**Comment by peppelinux** [comment]
I encountered 2 exceptions after merge your commits, 
please see https://github.com/peppelinux/djangosaml2idp/commit/c128678f58b6512679d6d10df02612b4e3ab103e

You should merge this into yours I think

**Pull request merged**
